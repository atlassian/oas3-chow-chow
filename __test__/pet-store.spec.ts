import ChowChow from '../src';
import ChowError, { RequestValidationError } from '../src/error';
const fixture = require('./fixtures/pet-store.json');

describe('Pet Store', () => {
  let chowchow: ChowChow;

  beforeAll(() => {
    chowchow = new ChowChow(fixture as any);
  })

  test('It should throw an error if a path is undefined', () => {
    expect(() => {
      chowchow.validateRequest('/undefined', {
        method: 'get'
      });
    }).toThrowError(ChowError);
  });

  test('It should successfully throw an error if a method is undefined', () => {
    expect(() => {
      chowchow.validateRequest('/pets', {
        method: 'put'
      })
    }).toThrowError(ChowError);
  })

  describe('Path', () => {
    test('It should fail validation if provided path parameter is wrong', () => {
      expect(() => {
        chowchow.validateRequest('/pets/chow', {
          method: 'get',
          path: {
            petId: 'chow'
          }
        });
      }).toThrowError(ChowError);
    })

    test('It should pass validation if provided path parameter is correct', () => {
      expect(() => {
        chowchow.validateRequest('/pets/123', {
          method: 'get',
          path: {
            petId: 123
          }
        });
      }).not.toThrowError();
    })
  })

  describe('Query', () => {
    test('It should fail validation if provided query parameter is wrong', () => {
      expect(() => {
        chowchow.validateRequest('/pets', {
          method: 'get',
          query: {
            limit: 'xyz'
          }
        });
      }).toThrowError(ChowError);
    })

    test('It should pass validation if provided path parameter is correct', () => {
      expect(() => {
        chowchow.validateRequest('/pets', {
          method: 'get',
          query: {
            limit: 50
          }
        });
      }).not.toThrowError();
    })

    test('It should pass validation if an array is passed to parameter which should be an array', () => {
      expect(() => {
        chowchow.validateRequest('/pets', {
          method: 'get',
          query: {
            breed: ['chowchow']
          }
        })
      }).not.toThrowError();
    })

    test('It should fail validation if invalid item is passed in enum', () => {
      expect(() => {
        chowchow.validateRequest('/pets', {
          method: 'get',
          query: {
            breed: ['nice dog']
          }
        })
      }).toThrowError(ChowError);
    })

    test('It should fail validation if number of items exceeds the limit', () => {
      expect(() => {
        chowchow.validateRequest('/pets', {
          method: 'get',
          query: {
            breed: ['chowchow', 'bichon', 'jack russell', 'labrador']
          }
        })
      }).toThrowError(ChowError);
    })

    test('It should pass validation for valid array parameter', () => {
      expect(() => {
        chowchow.validateRequest('/pets', {
          method: 'get',
          query: {
            breed: ['chowchow', 'bichon', 'labrador']
          }
        })
      }).not.toThrowError();
    })
  })
  describe('RequestBody', () => {
    test('It should fail validation if payload is invalid', () => {
      expect(() => {
        chowchow.validateRequest('/pets', {
          method: 'post',
          body: {
            name: 'plum',
          },
          header: {
            'content-type': 'application/json'
          }
        })
      }).toThrowError(ChowError)
    })

    test('It should fail validation if invalid mediaType is asked', () => {
      expect(() => {
        chowchow.validateRequest('/pets', {
          method: 'post',
          body: {
            id: 123,
            name: 'plum'
          },
          header: {
            'content-type': 'application/awsome'
          }
        })
      }).toThrowError(ChowError)
    })

    test('It should fail validation if requestBody is required but missing', () => {
      expect(() => {
        chowchow.validateRequest('/pets', {
          method: 'post',
          header: {
            'content-type': 'application/json'
          }
        })
      }).toThrowError(ChowError)
    })

    test('It should fail validation if requestBody is required but Content type is missing', () => {
      expect(() => {
        chowchow.validateRequest('/pets', {
          method: 'post',
        })
      }).toThrowError(ChowError)
    });

    test('It is ok to ignore body if it is not required', () => {
      expect(() => {
        chowchow.validateRequest('/pets/123', {
          method: 'post',
          path: {
            petId: 123
          },
          header: {
            'content-type': 'application/json'
          }
        })
      }).not.toThrowError()

    })

    test('It should pass validation if valid requestBody is passed', () => {
      expect(() => {
        chowchow.validateRequest('/pets', {
          method: 'post',
          body: {
            id: 123,
            name: 'plum',
            writeOnlyProp: '42',
            notReadOnlyProp: '42'
          },
          header: {
            'content-type': 'application/json'
          }
        })
      }).not.toThrowError()
    })

    test('It should fail validation if requestBody with readOnly property passed', () => {
      expect(() => {
        chowchow.validateRequest('/pets', {
          method: 'post',
          body: {
            id: 123,
            name: 'plum',
            readOnlyProp: '42'
          },
          header: {
            'content-type': 'application/json'
          }
        })
      }).toThrow(RequestValidationError)
    })
  })

  describe('Header', () => {
    test('It should fail validation if a required header is missing', () => {
      expect(() => {
        chowchow.validateRequest('/test/header', {
          method: 'get',
          header: {
            'content-type': 'application/json'
          }
        })
      }).toThrowError(ChowError)
    })

    test('It should fail validation if a header fails schema validation', () => {
      expect(() => {
        chowchow.validateRequest('/test/header', {
          method: 'get',
          header: {
            version: 'awsome version'
          }
        })
      }).toThrowError(ChowError)
    })

    test('It should pass validation if headers are satisfied', () => {
      expect(() => {
        chowchow.validateRequest('/test/header', {
          method: 'get',
          header: {
            version: 123
          }
        })
      }).not.toThrowError()
    })

    test('It should pass for header without schema', () => {
      expect(() => {
        chowchow.validateRequest('/test/header', {
          method: 'get',
          header: {
            version: 123,
            'no-schema': 123
          }
        })
      }).not.toThrowError()
    })
  })

  describe('Cookie', () => {
    test('It should fail validation if a required cookie is missing', () => {
      expect(() => {
        chowchow.validateRequest('/test/cookie', {
          method: 'get',
          cookie: {}
        })
      }).toThrowError(ChowError)
    })

    test('It should fail validation if a cookie fails schema validation', () => {
      expect(() => {
        chowchow.validateRequest('/test/cookie', {
          method: 'get',
          cookie: {
            count: 'many'
          }
        })
      }).toThrowError(ChowError)
    })

    test('It should pass validation if cookies are satisfied', () => {
      expect(() => {
        chowchow.validateRequest('/test/cookie', {
          method: 'get',
          cookie: {
            count: 123
          }
        })
      }).not.toThrowError()
    })
  })

  describe('Schema', () => {
    test('It is ok to not give a schema', () => {
      expect(() => {
        chowchow.validateRequest('/test/schema', {
          method: 'get',
          cookie: {
            count: 123
          }
        })
      }).not.toThrowError()
    })

    test('It is ok to use wildcards', () => {
      expect(() => {
        chowchow.validateRequest('/test/wildcard', {
          method: 'post',
          body: {
            id: 123,
            name: 'plum'
          },
          header: {
            'content-type': 'application/awesome'
          }
        })
      }).not.toThrowError()
    })

    test('It is ok to fall back to */* when no content type is provided', () => {
      expect(() => {
        chowchow.validateRequest('/test/wildcard', {
          method: 'post',
          body: [
            {
              id: 123,
              name: 'plum'
            },
            {
              id: 456,
              name: 'chow'
            }
          ]
        })
      }).not.toThrowError()
    })

  })

  describe('OperationId', () => {

    test('It should return unique operationId', () => {
      const validatedRequest = chowchow.validateRequest('/pets/123', {
        method: 'get',
        path: {
          petId: 123
        }
      });
      expect(validatedRequest.operationId).toEqual('showPetById');
    })

    test('It should respect custom operationId', () => {
      const validatedRequest = chowchow.validateRequest('/pets/123', {
        method: 'get',
        operationId: 'customId',
        path: {
          petId: 123
        }
      });
      expect(validatedRequest.operationId).toEqual('customId');
    })

  })

});
