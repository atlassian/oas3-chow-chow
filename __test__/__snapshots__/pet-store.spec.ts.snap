// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pet Store RequestBody It returns defined body content type 1`] = `
Array [
  "application/json",
]
`;

exports[`Pet Store RequestBody It returns empty array for defined body content type if method is undefined 1`] = `Array []`;

exports[`Pet Store RequestBody It returns empty array for defined body content type if path is undefined 1`] = `Array []`;

exports[`Pet Store RequestBody It returns empty array for defined body content type if requestBody is not defined 1`] = `Array []`;
